core/src/main/java/org/apache/iceberg/io/WriteBuilder.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import org.apache.iceberg.FileContent;
+import org.apache.iceberg.MetricsConfig;
+import org.apache.iceberg.Schema;
+
+/**
+ * Builder interface for creating file writers across supported data file formats. Each {@link
+ * FileAccessFactory} implementation provides appropriate {@link WriteBuilder} instances based on:
+ *
+ * <ul>
+ *   <li>target file format (Parquet, Avro, ORC)
+ *   <li>engine-specific object representation (spark, flink, generic, etc.)
+ *   <li>content type ({@link FileContent#DATA}, {@link FileContent#EQUALITY_DELETES}, {@link
```

Comment:
Why should the writer need to know the content type? Shouldn't it just need a schema?

Reply:
There are 2 main things:
- WriterContext - the writer needs the info if it is a data or a delete file we are writing. Based on this different table properties are used, like `write.delete.parquet.row-group-size-bytes`, or `write.parquet.row-group-size-bytes`
- The logic of writing the positional deletes are dependent on the file format `PositionDeleteStructWriter` for Parquet, `PositionAndRowDatumWriter` for Avro, `GenericOrcWriters.positionDelete` for ORC.  They need different handling for different file formats, which I wanted to keep internal to the file formats themselves. But then they need the info